## Introduction to Data Science

#### University of Redlands - DATA 101
#### Prof: Joanna Bieri [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
#### [Class Website: data101.joannabieri.com](https://joannabieri.com/data101.html)

---------------------------------------
# Homework Day 2
---------------------------------------

GOALS:

1. Start using Python
2. Start looking at Data!

----------------------------------------------------------

This homework has **TWO** warm-up problems and **FOUR** problems.


## Hello World!

### Warm up Problem 1

Write python code that will print your name and run the cell.

In [1]:
# Your code here
print('hello world')


hello world


## Installing Modules

### Warm up Problem 2

Install the modules that you will need to run the rest of the code. You can copy and paste the commands from the class notes. This might take a minute or two to run... go get a cup of coffee :)

In [1]:
# Copy and paste the code here:
!conda install -y numpy
!conda install -y pandas
!conda install -y matplotlib
!conda install -y plotly
!conda install -y itables
!conda install -y statsmodels
!conda install -y -c conda-forge python-kaleido

Retrieving notices: done
Channels:
 - conda-forge
 - defaults
Platform: win-64
Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\jfran\anaconda3\envs\Data101

  added / updated specs:
    - numpy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    icu-75.1                   |       he0c23c2_0        13.9 MB  conda-forge
    libblas-3.9.0              |  35_h5709861_mkl          64 KB  conda-forge
    libcblas-3.9.0             |  35_h2a3cdd5_mkl          65 KB  conda-forge
    libhwloc-2.12.1            |default_h64bd3f2_1002         2.3 MB  conda-forge
    libiconv-1.18              |       hc1393d2_2         681 KB  conda-forge
    liblapack-3.9.0            |  35_hf9ab0e9_mkl          77 KB  conda-forge
    libwinpthread-12.0.0.r4.gg4f2fc60ca|       h57928b3_9          35 KB  conda-forge
    libxml2-2.15.0             |       ha29bfb0_0       

## Let's do some Data Science.

We will explore some data about how countries vote at the United Nations General Assembly. You should follow along with the notes and/or class video.

I am breaking this into parts and having you copy and paste the code so you can start to identify what the different parts of the code do.

-------------------------------
### Importing Packages:
-------------------------------

Every time we start a new project we will import the packages that will help us do the analysis. Copy and paste all of the imports in the cell below.

* numpy = mathematical and number packages for python
* pandas = pretty tables, dataframes, and data analysis packages
* matplotlib.pyplot = nice looking graphs
* plotly = nice looking graphs.
* itables = pretty looking tables that have a search bar

In [8]:
# Copy and paste the code here:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.default='colab'

from itables import show

-------------------------------
### Get the Data:
-------------------------------

We will practice all sorts of ways to get data into our notebooks. When you use Pandas (pd) to read a csv:

    DF = pd.read_csv(file_location)

you are basically reading in a spreadsheet (Excel). The data for today is stored on my website.

It might take a minute for the data to load.

In [12]:
# Copy and paste the code here:
file_location = 'https://joannabieri.com/introdatascience/data/unvotes.csv'
DF = pd.read_csv(file_location)
years = [int(d.split('-')[0]) for d in DF['date']]
DF['year'] = years
DF = DF.drop('Unnamed: 0',axis=1)

show(DF)


-------------------------------
### Initial Data Exploration:
-------------------------------
-------------------------------
### Show the Data:
-------------------------------

We want to look at what we have. The data is now represented by the variable DF.

In [11]:
# Copy and paste the code here:
DF

rcid             country country_code     vote  session  \
0          6       United States           US       no        1   
1          6              Canada           CA       no        1   
2          6                Cuba           CU      yes        1   
3          6  Dominican Republic           DO  abstain        1   
4          6              Mexico           MX      yes        1   
...      ...                 ...          ...      ...      ...   
857873  9101           Venezuela           VE      yes       74   
857874  9101             Vietnam           VN      yes       74   
857875  9101               Yemen           YE      yes       74   
857876  9101              Zambia           ZM      yes       74   
857877  9101            Zimbabwe           ZW      yes       74   

        importantvote        date        unres  amend  para  \
0                 0.0  1946-01-04      R/1/107    0.0   0.0   
1                 0.0  1946-01-04      R/1/107    0.0   0.0   
2                 0.0  1946-01-04      R/1/107    0.0   0.0   
3                 0.0  1946-01-04      R/1/107    0.0   0.0   
4                 0.0  1946-01-04      R/1/107    0.0   0.0   
...               ...         ...          ...    ...   ...   
857873            NaN  2019-12-03  A/RES/74/12    NaN   NaN   
857874            NaN  2019-12-03  A/RES/74/12    NaN   NaN   
857875            NaN  2019-12-03  A/RES/74/12    NaN   NaN   
857876            NaN  2019-12-03  A/RES/74/12    NaN   NaN   
857877            NaN  2019-12-03  A/RES/74/12    NaN   NaN   

                                                    short  \
0                             DECLARATION OF HUMAN RIGHTS   
1                             DECLARATION OF HUMAN RIGHTS   
2                             DECLARATION OF HUMAN RIGHTS   
3                             DECLARATION OF HUMAN RIGHTS   
4                             DECLARATION OF HUMAN RIGHTS   
...                                                   ...   
857873  Division for Palestinian Rights of the Secreta...   
857874  Division for Palestinian Rights of the Secreta...   
857875  Division for Palestinian Rights of the Secreta...   
857876  Division for Palestinian Rights of the Secreta...   
857877  Division for Palestinian Rights of the Secreta...   

                                                    descr short_name  \
0       TO ADOPT A CUBAN PROPOSAL (A/3-C) THAT AN ITEM...         hr   
1       TO ADOPT A CUBAN PROPOSAL (A/3-C) THAT AN ITEM...         hr   
2       TO ADOPT A CUBAN PROPOSAL (A/3-C) THAT AN ITEM...         hr   
3       TO ADOPT A CUBAN PROPOSAL (A/3-C) THAT AN ITEM...         hr   
4       TO ADOPT A CUBAN PROPOSAL (A/3-C) THAT AN ITEM...         hr   
...                                                   ...        ...   
857873  Division for Palestinian Rights of the Secreta...         me   
857874  Division for Palestinian Rights of the Secreta...         me   
857875  Division for Palestinian Rights of the Secreta...         me   
857876  Division for Palestinian Rights of the Secreta...         me   
857877  Division for Palestinian Rights of the Secreta...         me   

                       issue  year  
0               Human rights  1946  
1               Human rights  1946  
2               Human rights  1946  
3               Human rights  1946  
4               Human rights  1946  
...                      ...   ...  
857873  Palestinian conflict  2019  
857874  Palestinian conflict  2019  
857875  Palestinian conflict  2019  
857876  Palestinian conflict  2019  
857877  Palestinian conflict  2019  

[857878 rows x 15 columns]

### Problem 1

Do some initial exploration of this data and write about it in the markdown cell below.

* How man columns are there?
* What do the columns represent?
* What countries are there?
* Can you search for a country you are interested in?
* How many rows are there?
* What other observations do you have?



**(Double Click Here)**
15 columns
They represent the options for whats inside the collums
200 countries
Put them into a unique list
command f? 
857878 
the data is a little hard to read and understand in the way that it was put into the table


-------------------------------
### Make Python Explore the Data!:
-------------------------------

Copy and paste the command that will show you a list of the counties:

In [16]:
# Copy and paste the code here:
# Python can list all the different countries:
country_list = list(DF['country'].unique())

# Show the data in a nice way
show(pd.DataFrame(country_list,columns=['country']))

Copy and paste the command that will count up the number of countries:

In [ ]:
# Copy and paste the code here:


### Problem 2

Try writing some code for your self. Above we found a countries list by focusing on the column named 'country'. See if you can redo that same code but change it to focus on the column named 'issue'.

What do I expect here:

* First copy and past the code from above
* Then change that code slightly
* Run the cell to see if it worked

In [34]:
# Write your code here
print(f'There arre {len(country_list)} in our data set!')


There arre 200 in our data set!


-------------------------------
### Data Visualization:
-------------------------------

Now we will select three countries that we are interested in and see how their votes have changed over time. Below you should see code that selects: Turkey, United States, and United Kingdom.

**IMPORTANT** These have to be spelled and capitalized exactly like they are in the data. Python is unforgiving of typos!

You can just run the cell below - assuming you have done all the parts above!

In [37]:
countries = ['Honduras', 'Canada', 'South Africa']
issues = list(DF['issue'].unique())
c_groups = DF.groupby(['country','issue'])
print(issues)

['Human rights', 'Economic development', 'Colonialism', 'Palestinian conflict', 'Arms control and disarmament', 'Nuclear weapons and nuclear material']


Now that we have our focus countries, we can make pretty pictures.

You can just run the cell below - assuming you have done all the parts above!

In [39]:
def make_plot(countries,issue):
    '''
    A Python function that takes in the list of countries and issues and makes
    a scatter plot of each issue with a trendline for each country.
    '''
    x_data = []
    y_data = []
    c_data = []
    for cntry in countries:
        my_group = c_groups.get_group((cntry,issue))
        for y in my_group['year'].unique():
            x_data.append(y)
            tot_yes = sum(my_group[my_group['year']==y]['vote']=='yes')
            percent_yes = tot_yes/len(my_group[my_group['year']==y])*100
            y_data.append(percent_yes)
            c_data.append(cntry)

    fig = px.scatter(x=x_data, y=y_data,color=c_data,trendline="lowess",labels={"color": "Country"})

    fig.update_layout(
        title={
            'text': issue + '<br>',
            'y':0.9,
            'x':0.5,
            'xanchor': 'center',
            'yanchor': 'top'})
    fig.update_yaxes(title_text="% Yes")
    fig.update_xaxes(title_text="Year")
    fig.show()
    
for iss in issues:
    make_plot(countries,iss)

### Problem 3

Choose one of the graphs produced above (should match the lecture notes) and discuss what you see there:

1. Which issue are you focusing on.
2. What do the graph axis represent?
3. Say in words how the United States votes have changed over time.
4. How do the United States votes compare to the other countries?
5. See if you can come up with one or two questions that interest you about the data in the graph.

#1. Focusing on weapons and material over the years
#2. Axis represents the amount of votes yes, and the years that the votes happened
#3. The United States votes had plateaued for a while before stedilly trending downward.
#4. Although another NATO countries, turkey and United Kingdom both are represented, only turkeys votes trend upwards, unlike the United States.
#5. What are the amounts of nuclear material that all three of these counrties posess?

### Problem 4

Now go back up to where we picked our three focus countries and choose some different ones. Rerun the analysis. If you want you could choose more than three countries. 

Discuss the graph for the same issue but with the new countries.

**(double click here)**

--------------------------------
### You are done with the homework... now what?
--------------------------------

1. Save your changes.
2. In the Git tab, **Stage** your changed files - use the (+) button next to the file.
3. **Commit** your changes by entering a summary and pushing commit.
4. **Push** your changes using the cloud button.
5. Check that your changes are on your GitHub repo (online)
6. **TAKE THE DAILY QUIZ ON CANVAS** the quiz will ask you to copy and past the HTTPS link to your repo for the day.
7. Come to class and get your questions answered.

At the end of the week  you will submit your final versions of the homework on Canvas for grading. For this you can just drag and drop the HW.ipynb files into canvas. 

**IMPORTANT** If this is confusing this first week, don't worry, I can help you and I will be really flexible about these first few deadlines.
